Script to reset ckcp

ramessesii2 opened the pull request:
#### This PR introduces a script to clean up the resources deployed/used by `ckcp`.

Signed-off-by: Satyam Bhardwaj <sabhardw@redhat.com>

bnallapeta commented on ckcp/hack/util/destroy.sh:
```diff
@@ -0,0 +1,96 @@
+
+    # remove namesapces generated by ckcp
+    printf "\n Removing namespaces created by ckcp:\n"
+    ckcp_generated_n="$(kubectl get ns | cut -d " " -f 1 | grep kcp)"
```
ckcp_generated_n  => ckcp_generated_ns

ramessesii2 replied:
done thanks

bnallapeta reviewed (changes requested):
Apart from the comments inline, I see that there's no error handling anywhere.. what happens when something does not get deleted? The script is not handling those scenarios.

bnallapeta commented on ckcp/hack/util/destroy.sh:
```diff
@@ -0,0 +1,96 @@
+
+    # remove namesapces generated by ckcp
+    printf "\n Removing namespaces created by ckcp:\n"
+    ckcp_generated_n="$(kubectl get ns | cut -d " " -f 1 | grep kcp)"
```
this command fetches all namespaces with substring 'kcp'. Would need to be changed to fetch ns starting with 'kcp-' as the generated namespaces follow this pattern.

```
[bnr@bnr pipelines-service]$ k get ns | cut -d " " -f1  | grep kcp
ckcp
kcp-1tfpg81ue6h0
kcp-mujo6gr3n42i
kcp-syncer-0d30625518f0617f253a43fc9339ea91-ziyfk3sn
kcp-zbbksthmbir1
```

ramessesii2 replied:
I've changed this part by first fetching a label which would be applied to all the namespaces created by the kcp-ws in the compute. And then selecting the NS by the label selector.

ramessesii2 commented on ckcp/hack/util/destroy.sh:
```diff
@@ -0,0 +1,96 @@
+# limitations under the License.
+
+SCRIPT_DIR="$(
+
```
Done

ramessesii2 commented on ckcp/hack/util/destroy.sh:
```diff
@@ -0,0 +1,96 @@
+  pwd
+)"
+
+CONFIG="$SCRIPT_DIR/../../config.yaml"
```
sure

ramessesii2 commented on ckcp/hack/util/destroy.sh:
```diff
@@ -0,0 +1,96 @@
+    # Remove Pipeline-Service application
+    kubectl delete -f "$SCRIPT_DIR/../../../gitops/argocd/argocd.yaml"
+    # Remove all the applications deployed by Pipeline-Service
+    kubectl delete -k "$SCRIPT_DIR/../../../gitops/argocd/argo-apps"
```
I checked it. Although they're deployed as child apps of Pipeline Service, they seem to have an independent existence as they're also deployed as an Argo CD application. That's why we would need to delete them separately.

ramessesii2 commented:
> Apart from the comments inline, I see that there's no error handling anywhere.. what happens when something does not get deleted? The script is not handling those scenarios.

That's true. Now, there are few checks around.
Moreover, I've removed uninstallation of various operators because that could lead to unintentional loss of user data/resources. 
If we feel like it's a good idea to have this option, we can use flags/arguments to trigger removal of various operators!

Roming22 commented:
There seems to be a few unresolved conversation. Let us know when the PR needs another review.

ramessesii2 commented on ckcp/hack/util/destroy.sh:
```diff
@@ -0,0 +1,96 @@
+    printf "\n Uninstalling Openshift-Pipelines Operator:\n"
+    kubectl delete tektonconfigs.operator.tekton.dev config
+    kubectl delete subscriptions.operators.coreos.com -n openshift-operators openshift-pipelines-operator
+    kubectl delete csv openshift-pipelines-operator-rh.v1.7.2 -n openshift-operators
```
done

ramessesii2 commented on ckcp/hack/util/destroy.sh:
```diff
@@ -0,0 +1,96 @@
+# Uninstalls operators
+uninstall_operators(){
+    # Openshift-Pipelines Operator
+    printf "\n Uninstalling Openshift-Pipelines Operator:\n"
```
Yes, I've tried making the operator uninstallation more generic. 
There's still few exceptions in case of Pipelines Operator for e.g.- we've to start with uninstalling TektonConfig and then the rest of the resources related to Pipelines Operator.
So, my the common steps in uninstalling operators are - 
1. Remove the subscription.
2. Remove csv
3. Delete all the related CRDs
4. Delete the operator (doing this before deleting the CRDs would just respawn the operator).

ramessesii2 commented:
Hi @Roming22 @bnallapeta, I've addressed all of the comments on this PR. Please let me know if you're okay with them and take a quick look at this PR again.
Few points -
1. For uninstalling operators and remove their traces from the cluster, this script will do force cleanup of CRDs by patching them.
2. This script wouldn't halt at errors(or engulf them) in case a resource meant to be deleted is already not available. This is for places where we're directly referencing files for e.g Syncer manifest or an Argo CD application referred.
3. It will halt at errors which would somehow stop the intended job of this script.

bnallapeta commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+    --reset-hard RESET_HARD
+        setting this option to 'true'
+        will aggresively remove operators deployed as well.
+        Defalut value: \$RESET_HARD
```
Default

Roming22 replied:
`Defalut value: \$RESET_HARD` -> `Default value: %s`, and add $RESET towards the end of the print.

bnallapeta commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+Optional arguments:
+    --reset-hard RESET_HARD
+        setting this option to 'true'
+        will aggresively remove operators deployed as well.
```
will **aggressively** remove operators that were installed by the openshisft_dev_setup.sh script.

bnallapeta commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+    fi
+
+    # remove syncer resources
+    printf "\n  Removing resources taken up by the syncer:\n"
```
created by the syncer*

bnallapeta commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+        break
+      fi
+    done
+    printf "\nckcp reset successfull.\n"
```
successful*

bnallapeta commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+  fi
+}
+
+# Removes different Argo CD applications deployed
```
# Removes Argo CD applications that were deployed by openshift_dev_setup.sh

bnallapeta commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+prechecks() {
+  KCP_KUBECONFIG=${KCP_KUBECONFIG:-}
+  if [[ -z "${KCP_KUBECONFIG}" ]]; then
+    printf "\n\n[ERROR] Missing parameter --kcp-kubeconfig"
```
printf "\n[ERROR] Missing parameter --kcp-kubeconfig\n"

Roming22 replied:
Also, please redirect to stderr with `>&2`. Occurs multiple times in the file.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+Scrap ckcp and free resources deployed by openshift_dev_setup.sh script.
+
+Mandatory arguments:
+    --kcp-kubeconfig KCP_KUBECONFIG
```
I would not use `--kcp-kubeconfig` and `--kubeconfig`, but instead use `--work-dir`, to mirror the use of the `openshift_dev_setup.sh` script.
I think it's fair to assume that the script is not expected to work on a environment that was not set by `openshift_dev_setup.sh`.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+        Display this message.
+
+Example:
+    %s --kcp-kubeconfig './tmp/tmp.435kjkdsf/credentials/kubeconfig/kcp/ckcp.kubeconfig'
```
If you change to work-dir, remember to change the example.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+  KCP_KUBECONFIG=${KCP_KUBECONFIG:-}
+  if [[ -z "${KCP_KUBECONFIG}" ]]; then
+    printf "\n\n[ERROR] Missing parameter --kcp-kubeconfig"
+    exit
```
`exit 1`, otherwise it fails silently for the shell. Occurs multiple times in the file.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+
+prechecks() {
+  KCP_KUBECONFIG=${KCP_KUBECONFIG:-}
+  if [[ -z "${KCP_KUBECONFIG}" ]]; then
```
Might need to be changed to "$WORK_DIR"

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+    # list of all the Argo Apps still deployed
+    mapfile -t argo_apps_deployed < <(argocd app list -o yaml | yq '.[].metadata.name')
+    matched_apps=()
+    for app in "${all_argo_apps[@]}"
```
Please favor the `for ...; do` syntax (single line statement). Occurs multiple times in the file.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+    # created in the compute cluster by the same kcp-workspace.
+    ns_label=$(KUBECONFIG=${KCP_KUBECONFIG} kubectl get synctargets.workload.kcp.dev -o yaml | yq '.items[0].metadata.labels' | grep -ie "internal.workload.kcp.dev/key")
+    if [[ -z $ns_label ]]; then
+        printf "\n\n[ERROR] Couldn't extract labels from kcp synctarget object.\n"
```
Redirect the `ERROR` message to stderr with `>&2`, and use `exit 1`.
The helpful message can stay on stdout.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+
+    # Check if the Argo CD applications have been indeed removed
+    # list of all the Argo Apps that a user can deploy using Pipeline Service
+    mapfile -t all_argo_apps < <(kubectl kustomize "$GITOPS_DIR/argocd/argo-apps"  | yq '.metadata.name' | grep -e '[a-z]')
```
`grep -v '^---$'` is a bit more descriptive.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+    then
+      for crd in "${argo_crds[@]}"
+      do
+        kubectl delete crd "$crd" &
```
Why is it run in the background? If you patched the resource first to remove the finalizer, and then deleted it, shouldn't it work in the foreground?

If this requires a change, this appears multiple times in the file.

ramessesii2 replied:
As discussed, the patch operation before deleting(i.e., resource is in terminating state) is not allowed. So, we'll keep it in favor of removing(forcefully) CRDs.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+    ns_label=$(KUBECONFIG=${KCP_KUBECONFIG} kubectl get synctargets.workload.kcp.dev -o yaml | yq '.items[0].metadata.labels' | grep -ie "internal.workload.kcp.dev/key")
+    if [[ -z $ns_label ]]; then
+        printf "\n\n[ERROR] Couldn't extract labels from kcp synctarget object.\n"
+        printf "Make sure you pass the correct kcp_kubeconfig directory using --kcp_kubeconfig arg"
```
Add `\n` at the end of the string.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+    # remove syncer resources
+    printf "\n  Removing resources taken up by the syncer:\n"
+    current_context="$(yq e ".current-context" <"$KUBECONFIG")"
+    current_context=$(echo "$current_context" | sed  's,default/,,g; s,:6443/kube:admin,,g')
```
Double space after `sed`

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+    printf "\n  Removing resources taken up by the syncer:\n"
+    current_context="$(yq e ".current-context" <"$KUBECONFIG")"
+    current_context=$(echo "$current_context" | sed  's,default/,,g; s,:6443/kube:admin,,g')
+    kubectl delete -f /tmp/syncer-"$current_context".yaml
```
Where does this file comes from? What happens if I reboot my laptop and /tmp is cleaned? I would avoid on using resources that are not in the work dir or not in the repo.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+    mapfile -t ckcp_generated_ns < <(kubectl get ns -l "${ns_label//\/key: /\/cluster=}" | cut -d " " -f 1 | tail -n +2)
+    for ns in "${ckcp_generated_ns[@]}"
+    do
+        echo " - $ns"
```
`printf "  - %s\n" "$ns"` for consistency.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,288 @@
+      if (( i == 7 )); then
+        printf "\n\n[ERROR] Couldn't remove the namespaces created by ckcp: %s ${ckcp_generated_ns[*]} "
+        exit
+      fi
```
You should be able to replace the `fi; if` with `elif`.

Roming22 commented on ckcp/README.md:
````diff
@@ -112,3 +112,10 @@ Considerations for testing a new component:-
+
+One can reset `ckcp` and all the resources deployed by it using:-
+```bash
+./ckcp/hack/util/reset_ckcp.sh --kcp-kubeconfig=/to/kcp_kubeconfig
````
That should now mention `--work-dir` instead.

ramessesii2 replied:
ah, missed that.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,289 @@
+  pwd
+)"
+
+PLNSVC_DIR="$(dirname "$(dirname "$(dirname "$SCRIPT_DIR")")")"
```
`PLNSVC_DIR` is usually called `PROJECT_DIR` in other scripts.
I would simplify to `"$(cd "$SCRIPT_DIR/../../.." >/dev/null || exit 1; pwd)"`

bnallapeta replied:
@Roming22 a while back, one of the jobs was flaky because CI was unable to check ../.. structure. That's why I feel its better to use dirname.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,289 @@
+        Kubeconfig files for compute clusters are expected in the subdirectory: credentials/kubeconfig/compute
+
+Optional arguments:
+    --reset-hard RESET_HARD
```
The argument seems unnecessary, the option seems to be a flag.

Roming22 commented on ckcp/hack/util/reset_ckcp.sh:
```diff
@@ -0,0 +1,289 @@
+    fi
+    KUBECONFIG="$WORK_DIR/credentials/kubeconfig/compute/compute.kubeconfig.base"
+    if [ ! -f "$KUBECONFIG" ]; then
+      printf "\n[ERROR] Couldn't find compute's kubeconfig." >&2
```
It's a good idea to print the expected file path to give users a starting point for debugging. Applicable for the kcp kubeconfig as well.

ramessesii2 replied:
sure

Roming22 commented:
@bnallapeta Feel free to merge if you do not have any more feedback.

bnallapeta reviewed (approved):
Tested it again and it works. Merging the PR.
